x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/ShardMultiSearchAction.java

```diff
@@ -0,0 +1,244 @@
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.text.Text;
+import org.elasticsearch.common.xcontent.DeprecationHandler;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.index.IndexService;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.fieldvisitor.FieldsVisitor;
+import org.elasticsearch.index.mapper.MapperService;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryShardContext;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.SearchHits;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.search.fetch.subphase.FetchSourceContext;
+import org.elasticsearch.search.internal.InternalSearchResponse;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.enrich.EnrichPolicy;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This is an internal action, that executes msearch requests for enrich indices in a more efficient manner.
+ * Currently each search request inside a msearch request is executed as a separate search. If many search requests
+ * are targeted to the same shards then there is quite some overhead in executing each search request as a separate
+ * search (multiple search contexts, opening of multiple searchers).
+ *
+ * In case for the enrich processor, searches are always targeting the same single shard indices. This action
+ * handles multi search requests targeting enrich indices more efficiently by executing them in a bulk using the same
+ * searcher and query shard context.
+ *
+ * This action (plus some coordination logic in {@link CoordinatorProxyAction}) can be removed when msearch can
+ * execute search requests targeted to the same shard more efficiently in a bulk like style.
+ *
+ * Note that this 'msearch' implementation only supports executing a query, pagination and source filtering.
+ * Other search features are not supported, because the enrich processor isn't using these search features.
+ */
+public class ShardMultiSearchAction extends ActionType<MultiSearchResponse> {
```

Comment:
should this carry Enrich in the name ?  I understand it _could_ be general purpose, but some of the validation and assumptions are Enrich specific.

Reply:
No, this code isn't general purpose. The idea behind it is. 
I just didn't use the `Enrich` prefix because is is already in the enrich package.
I will add the enrich prefix.

Reply:
i pushed: 54498ba2c761d8f383fe74e83d6bcefadf20c657